### ISSUE TITLE
Calling `addBreadcrumb` triggers `Main Thread Checker` warning

### ISSUE THREAD

**Issue by ncltg** [getsentry/sentry-cocoa#4015]
### Platform

iOS

### Environment

Develop

### Installed

Swift Package Manager

### Version

8.26.0

### Xcode Version

15.2

### Did it work on previous versions?

not sure

### Steps to Reproduce

Call `SentrySDK.addBreadcrumb()` from a background thread

### Expected Result

No main thread checker issue is found.

### Result
Xcode shows a Main Thread checker warning. The SDK is calling UIKit methods that require Main Thread execution on a background thread.

### Logs and screenshots

```
Main Thread Checker: UI API called on a background thread: -[UIWindow screen]
PID: 45981, TID: 2549765, Thread name: (none), Queue name: com.apple.root.user-initiated-qos.cooperative, QoS: 25
```

<img width="343" alt="Screenshot 2024-05-24 at 13 03 51" src="https://github.com/getsentry/sentry-cocoa/assets/4721089/7fdc7ab7-ed50-4c27-9725-9193268cf5bc">
<img width="599" alt="Screenshot 2024-05-24 at 13 03 42" src="https://github.com/getsentry/sentry-cocoa/assets/4721089/b5941983-dc8c-4551-93d5-8b1ce8bdc469">

### Question
I checked the iOS documentation and it does not mention that breadcrumb should be added from the main thread.
Is it an issue from Sentry SDK that needs to be addressed or does it actually recommended to call that method from the main thread ?

**Comment by ncltg** [comment]
Hey @brustolin 

> I believe this SentryLogRepository is an auxiliar class in your project correct?
Yes 👍

Based on the information you provided it's very likely that some of our launch tasks produce calls to addBreadcrumb before we start `SentrySDK` ourselves.

Thanks a lot for the clarification 🙏

**Comment by swhitty** [comment]
I also see main thread checker issues within `enrichScrope:` [8.26.0](https://github.com/getsentry/sentry-cocoa/releases/tag/8.26.0) but only _after_ calling `SentrySDK.close()`.
 

<img width="1966" alt="crash" src="https://github.com/getsentry/sentry-cocoa/assets/867662/b6d40b09-5420-4429-aec9-7bd63f5648c5">

**Comment by brustolin** [comment]
Hello @swhitty, thanks for reaching out.
By calling `SentrySDK.close()` you're cleaning up Sentry hub and its scope, by calling any function of the SDK a new one will be created.

But we need to improve this to not trigger a thread checker.

**Comment by philipphofmann** [comment]
We can change this to initialize a new SentryNoOpHub when there is none instead of creating a normal SentryHub. It's best to do this in the next major to avoid unwanted side effect.

**Comment by xinsight** [comment]
I'm seeing the same warning `Main Thread Checker: UI API called on a background thread: -[UIWindow screen]` after a network request. I'm not calling Sentry at all here.

Sentry 8.36.0

![Image](https://github.com/user-attachments/assets/9d1c57aa-7006-47b7-b3d4-2aef877b1332)

**Comment by philipphofmann** [comment]
Thanks for the update, @xinsight. I think we must cache the `screen_height_pixels` and `screen_width_pixels` and not retrieve them every time from the UIApplication to avoid accessing that on a BG thread
https://github.com/getsentry/sentry-cocoa/blob/7cb89e0cbd28986c4ed3c10b41318d1989bf2abf/Sources/Sentry/SentryCrashWrapper.m#L188-L195

This gets called frequently in the hub here
https://github.com/getsentry/sentry-cocoa/blob/7cb89e0cbd28986c4ed3c10b41318d1989bf2abf/Sources/Sentry/SentryHub.m#L519-L535

**Comment by brustolin** [comment]
> Thanks for the update, @xinsight. I think we must cache the screen_height_pixels and screen_width_pixels and not retrieve them every time from the UIApplication to avoid accessing that on a BG thread

Just need to remember that screen can change size with rotation, or iPad apps.

**Comment by philipphofmann** [comment]
> Just need to remember that screen can change size with rotation, or iPad apps.

Good point; we need to subscribe to `UIDeviceOrientationDidChangeNotification.`

**Comment by AnthonyOliveri** [comment]
To add another data point: I see this same Main Thread Checker warning when calling `SentrySDK.setUser(_)` if `SentrySDK.start()` is not called beforehand. I know that's a programming error, but may be helpful to know when fixing the issue.

**Comment by philipphofmann** [comment]
Thanks for the hint @AnthonyOliveri 👍

**Comment by philprime** [comment]
I did an investigation to understand what currently happens when `SentrySDK.addBreadcrumb(_:)` is used before the SDK is ready and **it actually returns early** because no options are available, not recording the breadcrumb anyways:

<img width="1049" height="390" alt="Image" src="https://github.com/user-attachments/assets/a01d170e-1da4-41b7-8734-ca3c6e8ba8a7" />

The underlying problem here is that the `enrichScope` is called when the hub is accessed before the SDK is initialized, which can also happen in background threads. **That's why we recommend start the Sentry SDK as soon as possible**, but it could still happen due to race-conditions.

Changing the `enrichScope` to be dispatched only on the main thread is a larger undertaking, so that will be tracked in #5424. 

As a reference, the same issue occurs when using `SentrySDK.setUser(_:)` before the SDK has been initialized. For `setUser` we decided to display a fatal error message to the user, so they are aware, but we don't stop the invalid SDK usage before initialization.

https://github.com/getsentry/sentry-cocoa/blob/3b373bb993db1aec3d39b8eaa5d43aaa70c5910f/Sources/Sentry/SentrySDKInternal.m#L511-L524

I see two options now:

1. Add a similar warning like `setUser` to explain and educate what's happening, without changing the logic of `addBreadcrumb` at all. This is non-breaking.
2. We add a guard-like statement to only add the breadcrumb if `SentrySDKInternal.isEnabled` is set to true (= the SDK is ready), basically blocking the method if the SDK is not ready. This is kinda breaking, but actually isn't because the breadcrumb isn't added anyways.

@philipphofmann @itaybre @kahest I would propose we go for number 1 to be consistent our existing approach, improve `enrichScope` in #5424 and in SDK v10+ we actually enforce that the SDK can only be used after it has been initialized and all public methods are thread-safe (open for discussion in another issue).

WDYT?

**Comment by kahest** [comment]
@philprime thanks for the analysis and the options - your suggestion sounds fair to me!

**Comment by philippsuess** [comment]
Thanks for the analysis. I also vote for option 1 including improve enrichScope in https://github.com/getsentry/sentry-cocoa/issues/5424.

> SDK v10+ we actually enforce that the SDK can only be used after it has been initialized

I think we could already now safeguard all calls to SentrySDK and log errors if it's not started yet, and yes all public methods should be thread safe of SentrySDK. These are two seperate discussions, I agree and we could open extra issues for these.
